### PR TITLE
added new serve command

### PR DESCRIPTION
**Pull request by kspearrin**
Asana task: https://app.asana.com/0/1153292148278596/1179203949573812

This PR adds a new command to the CLI: `bw serve`.

The `serve` command allows you to serve a local webserver (express) interface to the various CLI commands. This is useful for several reasons, including:

1. The CLI dependency chain is initialized once, therefore, subsequent commands should be faster.
2. You can call the CLI commands from an HTTP interface, in a RESTful-like way.

For example, you would normally call the `list` command like so:

```
bw list items
bw get item {GUID}
```

With the `serve` approach, you can call the same command via HTTP like so:

```
bw serve --port 8000

GET http://localhost:8000/list/items
GET http://localhost:8000/item/{GUID}
```

Unlocking the vault for `serve` HTTP requests can be done by passing a `--session` key to the serve command, such as:

```
bw serve --session {KEY}
```

Or, by passing the session key as a header with your HTTP requests:

```
GET http://localhost:8000/list/items

--Headers--
Session: {KEY}
```

### CODE REVIEW

**Review by MGibson1 — CHANGES_REQUESTED** [review]
I have two thoughts on this:

First, I don't want to add `any` as an option to our commands. I know commander's option we're passing in is already an `any`, but internal to our code I'll like to maintain type safety -- otherwise we may as well swap to js.

Luckily, express has a [`get` generic that specifies a parameters type](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/43434#issuecomment-607181516). We'd need to add a type definition with all the allowed options to each command.

Second, I'm also a little concerned with error handling in this. My gut is that we should error on unknown parameters or on parameters of the wrong type. 

 I think we can do that by either requiring a query params validator on every command object or by getting a little clever and defining a custom query parser. My gut is that would probably result in a similar params validator method on each command, but maybe we could do it based on types.
 
I think both of these concerns are solved by doing a little pre-processing of the `req.query` object. We'd validate it, and returned a typed options generated from the query parameters to be passed on to run.

**Comment by kspearrin** [comment]
@MGibson1 This should be good to review again now. I am now processing all known options passed to each command to ensure availability and the expected types. Error message have also been adjusted to be more generic for both CLI and HTTP/serve use-cases. Finally, I also added support for the remaining `send` commands.

**Comment by MGibson1** [comment]
@kspearrin reviewing, but build is failing due to too many parameters in the send create command in `serve`

File: src/program.ts [hunk 1]
```diff
@@ -395,6 +396,21 @@ export class Program extends BaseProgram {
...
+            .option('--port <port>', 'Provides a custom web vault URL that differs from the base URL.')
+            .on('--help', () => {
+                writeLn('\n  Examples:');
+                writeLn('');
+                writeLn('    bw serve');
+                writeLn('    bw serve --port 8080');
+                writeLn('', true);
+            })
+            .action(async cmd => {
+                const command = new ServeCommand(this.main);
```
[the remark this answers is not in the document]
Comment (kspearrin) [reply src/program.ts:411]: I dont think so, but it should probably have `exitIfNotAuthed`

[the remark this answers is not in the document]
Comment (MGibson1) [reply src/program.ts:411]: If no cryptoKey is present, most of the commands won't work. I suppose we'd throw in those cases, but there's no way to unlock an already running `Serve` command, so I'd think it's mostly not useful, right? At the least I'd like a warning that the vault is locked and crypto functions will be unavailable.

[the remark this answers is not in the document]
Comment (kspearrin) [reply src/program.ts:411]: @MGibson1 One of the features of serve is the ability to pass the session key in a request header, so there is definitely use case for running serve while locked.

[the remark this answers is not in the document]
Comment (MGibson1) [reply src/program.ts:411]: ah, I see, that does make sense

**Comment by kspearrin** [comment]
re-do here since all the breaking changes on the merge now. was easier just to re-do it. https://github.com/bitwarden/cli/pull/451

**Pull request closed without merging**